README.md
# Sign In Helper

⭐️[配套前端](https://github.com/rroy233/signin-helper-frontend)   🌈  [接口文档](/docs/devDoc.v1.md)  ☕️  [更新日志](CHANGELOG.md)

> 一个为班级成员提供打卡统计和打卡提醒服务的程序

![cover](docs/images/cover.jpg)

------



## 解决了什么问题

### 对于班级管理者

- 便于统计班级活动参与状况
- 便于文件（如截图、作业等）的收集，收齐后直接打包下载
- 提升班级活动趣味性、班级成员参与感

### 对于班级成员

- 若未完成任务，系统每日将发送提醒。对于一些时间跨度长的活动，避免出现以为自己完成了任务但实际却未完成的情况

### 对比

|              | 😢传统做法                                                    | 🥳使用本平台                                                  |
| ------------ | ------------------------------------------------------------ | ------------------------------------------------------------ |
| 统计完成情况 | 📝按照名单比对/🗣挨个询问                                      | 📊参与情况一目了然                                            |
| 每日提醒     | (❌不现实)                                                    | 邮件/微信 + 中午/晚上 提醒 ⏰                                 |
| 收集文件     | 使用小而美软件，文件缓存不清理，越用越臃肿💢。负责人逐个下载，整理文件麻烦 🤯 | 成员网页上传，管理员打包下载，文件自动命名，整个过程干干净净🤓 |


## 功能

- [x] 活动
  - [x] 多活动并行
  - [x] 自定义活动名称、封面图（自行上传/pixiv随机抓图）、简介等
  - [x] 设定是否需要上传文件
  - [x] 设定活动上传文件的种类、大小等
  - [x] 设定活动结束时间
  - [x] 设定是否开启每日提醒
  - [x] 预览/打包下载用户上传的文件
  - [x] 查看活动参与情况
  - [x] 手动开启/结束活动
  - [x] 临近截止时间，若参与率未达标将给管理员发送提醒
- [x] 用户
  - [x] 创建/加入/退出班级
  - [x] 查看活动参与记录
  - [x] 选择每日提醒推送方式
  - [x] 浏览自己上传的文件
  - [x] 管理员权限/普通用户权限区分，可变更身份
- [x] 通知
  - [x] 邮件推送
  - [x] 微信推送（使用第三方平台接口）
- [x] 文件
  - [x] pixiv抓取的文件（可能有涩图，所以）存储于程序目录
  - [x] 用户上传的文件存储于腾讯云对象存储（cos）
  - [x] 文件代理功能。所有文件访问请求均由`fileHandler`转发/加载
  - [x] 支持格式:jpg,png,pdf,zip,rar等
  - [x] 所有文件均有有效期，过期自动清除
- [x] 其他
  - [ ] 独立的登录/注册功能
  - [x] 推送内含有快捷登录的链接
  - [x] 随机推送模板
  - [x] 用Redis缓存部分计算结果，减轻高并发情况下服务器负载
  - [x] 用户登录状态可吊销



##  更多截图

|                          | 截图                                |
| ------------------------ | ----------------------------------- |
| 用户面板                 | ![image01](docs/images/image01.jpg) |
| 管理员面板               | ![image02](docs/images/image02.jpg) |
| 创建活动                 | ![image03](docs/images/image03.jpg) |
| 统计信息（无需上传文件） | ![image04](docs/images/image04.jpg) |
| 统计信息（需上传文件）   | ![image05](docs/images/image05.jpg) |


## 申请试用

由于该项目为练手项目，目前也未配备独立的登录注册功能，所以自己直接部署将无法正常使用。

如果需要试用请联系Email:oisroy233艾特gmail.com


CHANGELOG.md
# 更新日志

## [1.7.12] - 2022-09-10

### 修复

* 某管理员接口权限漏洞

## [1.7.11] - 2022-08-20

### 修复

* 登录漏洞

## [1.7.10] - 2022-07-16

### 修复

* 第三方随机图片接口

## [1.7.9] - 2022-07-03

### 变更

* 单点登录

## [1.7.8] - 2022-06-19

### 新增

* 静态资源缓存控制

## [1.7.7] - 2022-06-09

### 变更

* 不再为微信推送签发快捷登录链接

## [1.7.6] - 2022-05-22

### 新增

* 第三方统计

## [1.7.5] - 2022-04-17

### 优化

* 邮件推送正文使用短连接
* 随机图片获取

## [1.7.4] - 2022-04-14

### 变更

* 第三方随机图片接口

## [1.7.3] - 2022-03-01

### 优化

* 用户上传图片加载速度

## [1.7.2] - 2022-02-28

### 修复

* 由于第三方接口限制导致wx推送无法送达的bug



## [1.7.1] - 2022-02-20

### 变更

* actToken校验机制

### 修复

* 文件代理存在的漏洞

## [1.7.0] - 2022-01-28

### 新增

* 新增/移除管理员功能
* 班级成员管理显示邮箱
* 失效文件提示

### 修复

* 部分情况下管理员无法更改活动结束时间的问题
* 活动头图地址处理的部分bug

### 优化

* 初始化部分逻辑
* 其他



## [1.6.10]  - 2022-01-25

### 修复

* 微信推送的登录链接无效的问题

## [1.6.9]  - 2022-01-25

### 变更

* 加密快捷登录凭证


## [1.6.8]  - 2022-01-24

### 修复

* 部分zip文件无法通过类型检验的问题

### 新增

* 用户端活动列表分页

## [1.6.7] - 2022-01-12

### 修复

* 初始化账号部分逻辑错误

## [1.6.6] - 2022-01-08

### 新增

* 随机图片



## [1.6.5] - 2022-01-04

### 新增

* 过期文件清理

### 修复

* 某些缓存无法自动清理的问题



## [1.6.4] - 2022-01-01

### 新增

* 文件代理



## [1.6.2] - 2021-12-18

### 修复

* 上传文件时由于空指针出现`500错误`

## [1.6.1] - 2021-12-10

### 新增

* 管理员查看用户上传的文件
* 支持pdf



## [1.6.0] - 2021-12-05

### 新增

* 支持多种上传文件类型



## [1.5.3] -2021-12-03

### 优化

* 批量导出的部分逻辑



## [1.5.2] - 2021-12-02

### 新增

* 用户端预览文件



## [1.5.1] - 2021-12-01

### 修复

* 文件上传相关缓存未清理的问题

### 优化

* 用户页面管理部分样式和逻辑



### [1.5.0] - 2021-11-30

### 新增

* 用户管理
* 上传文件
* 活动的提醒开关

## [1.4.1] - 2021-11-21

## 修复

* 活动信息缓存回源查询过程存在的bug
* 手动关闭活动时，活动状态缓存更新不及时的bug

## 新增

* 在不变更活动状态时可以修改活动其他信息


## [1.4.0] - 2021-11-19

## 新增

* 调试功能

## [1.3.1] - 2021-11-15

## 修复

* 从快捷入口进入系统后出现"csrf-token mismatch"的错误

## [1.3.0] - 2021-11-13

### 新增

* 站内通知

### 变更

* actToken 生成算法



## [1.2.0] - 2021-11-10

### 新增

* 错误提示页面
* 防御机制

### 变更

* 管理员页面活动管理历史活动排序方式

### 修复

* 用户在初始化后仍然可以访问初始化页面



## [1.1.3] - 2021-11-7

### 新增

* 活动时间描述



## [1.1.2] - 2021-11-6

### 修复

* 图标无法显示的问题



## [1.1.1] - 2021-11-6

### 变更

* 更改日志记录方式



## [1.1.0] - 2021-11-5

### 变更

* 前端使用vue-cli重构



## [1.0.6] - 2021-11-4

### 新增

* 错误提示页
* 绑定微信后自动将微信设置为账号的通知方式

### 变更

* 微信推送细节
* 模板展示页仅对用户可见



## [1.0.4] - 2021-11-3

### 新增

* 细化通知机制，在活动即将结束时提醒管理员

### 变更

* 模板展示页面描述



## [1.0.3] - 2021-11-2

### 变更

* 将github链接替换为镜像站的，方便用户进入。



## [1.0.2] - 2021-11-1

### 变更

* 统计数据的排列顺序

### 修复

* 推送通知时可能导致的空指针问题



## [1.0.1] - 2021-10-31

### 新增

* 签到卡片渐变效果

### 修复

* 编辑活动时会导致图片丢失

### 变更

* 管理员端「活动管理」页面内，活动的排列顺序。



## [1.0.0] - 2021-10-29

### 新增

* 支持多任务并行，可以同时布置多个任务了

  

## [0.2.4] - 2021-10-28

### 新增

* 微信推送欢迎信息



## [0.2.3] - 2021-10-28

### 修复

* 中间件处理未授权请求的机制，解决了登录凭证失效后重新登录会重定向至Api的问题。



### 新增

* 支持微信推送



## [0.2.2] - 2021-10-27

### 变更

* 通知时间改为“中午12：30”和“下午06：30”



### 新增

* 支持每日邮件通知
* 支持管理员新建活动时群发通知
* 通知消息模板



## [0.1.0] - 2021-10-25

### 变更

* 首页排序方式，由按user_id排序改为按签到时间排序
* 通知更改页面描述



### 新增

* 显示版本号
* 管理员页面部分输入提示



## [0.0.0] - 2021-10-24

第一个版本


job.go
package main

import (
	"encoding/json"
	"github.com/robfig/cron/v3"
	"math/rand"
	"os"
	"signin/Logger"
	"strconv"
	"time"
)

var crontab *cron.Cron

func initJob() {
	crontab = cron.New(cron.WithLocation(TZ))
	var err error

	if config.General.Production == false {
		//debug
		_, err = crontab.AddFunc("15 19 * * ?", PrepareDailyNotification)
		_, err = crontab.AddFunc("16 19 * * ?", SendDailyNotification)
		if err != nil {
			Logger.FATAL.Fatalln("[定时任务][异常]添加Debugger成功:", err)
		} else {
			Logger.Info.Println("[定时任务][成功]添加Debugger成功")
		}
	}

	//清除已过期文件
	//每天凌晨1点
	_, err = crontab.AddFunc("0 0 * * ?", CleanExpiredFiles)
	if err != nil {
		Logger.FATAL.Fatalln("[定时任务][异常]添加CleanExpiredFiles失败:", err)
	} else {
		Logger.Info.Println("[定时任务][成功]添加CleanExpiredFiles成功")
	}

	//数据准备
	//每天12:29+18:29
	_, err = crontab.AddFunc("29 12 * * ?", PrepareDailyNotification)
	_, err = crontab.AddFunc("29 18 * * ?", PrepareDailyNotification)
	if err != nil {
		Logger.FATAL.Fatalln("[定时任务][异常]添加PrepareDailyNotification失败:", err)
	} else {
		Logger.Info.Println("[定时任务][成功]添加PrepareDailyNotification成功")
	}

	//发送任务
	//每天12:30+19:30
	_, err = crontab.AddFunc("30 12 * * ?", SendDailyNotification)
	_, err = crontab.AddFunc("30 18 * * ?", SendDailyNotification)
	if err != nil {
		Logger.FATAL.Fatalln("[定时任务][异常]添加SendDailyNotification失败:", err)
	} else {
		Logger.Info.Println("[定时任务][成功]添加SendDailyNotification成功")
	}

	crontab.Start()

}

func PrepareDailyNotification() {

	classList := make([]int, 0)
	err := db.Select(&classList, "SELECT  `class_id` FROM `class`")
	if err != nil {
		Logger.Error.Println("[定时任务][准备数据]获取班级列表失败", err)
	}

	//遍历每个班级
	for i, classID := range classList {
		Logger.Info.Println("[定时任务][准备数据]开始处理班级ID:", classID, "(", i+1, "/", len(classList), ")")

		//获取班级内所有用户
		users := make([]dbUser, 0)
		err = db.Select(&users, "select * from `user` where `class`=?", classID)
		if err != nil {
			Logger.Error.Println("[定时任务][准备数据]读取user表失败", err)
		}
		if len(users) == 0 {
			//太惨了，都关闭通知
			Logger.Info.Println("[定时任务][准备数据]开始处理班级ID:", classID, "无可发送用户")
			continue
		}

		//获取班级信息
		class, err := getClass(classID)
		if err != nil {
			Logger.Info.Println("[定时任务][准备数据]获取班级信息失败:", classID, err)
			continue
		}

		//获取班级有效活动
		actIDs, err := getActIDs(classID)
		if err != nil {
			Logger.Info.Println("[定时任务][准备数据]获取班级有效活动失败:", classID, err)
			continue
		}

		//遍历班级的每一个有效活动
		for _, actID := range actIDs {
			//获取参与数据
			sts, err := cacheActStatistics(actID)
			if err != nil {
				Logger.Info.Println("[定时任务][准备数据]获取参与数据失败")
				continue
			}

			act, err := getAct(actID)
			if err != nil {
				Logger.Info.Println("[定时任务][准备数据]获取活动信息失败:", actID, err)
				continue
			}

			//判断是否关闭每日提醒
			if act.DailyNotiEnabled == 0 {
				continue
			}

			//若全员完成，则不发送
			if sts.Done == sts.Total {
				Logger.Info.Println("[定时任务][准备数据]班级ID:", classID, "全员完成(", i+1, "/", len(classList), ")")
				continue
			}

			//id->user映射
			userMap := make(map[int]*dbUser, 0)
			for i := range users {
				userMap[users[i].UserID] = &users[i]
			}

			//遍历班级内所有未完成的同学
			for i := range sts.UnfinishedList {
				thisUser := userMap[sts.UnfinishedList[i].UserID]
				job, err := getTemplate(TPL_MSGTYPE_daily, TPL_LEVEL_LOW)
				if err != nil {
					Logger.Error.Println("[定时任务][准备数据]获取模板失败", err, thisUser)
					continue
				}
				var msgJson []byte
				//判断通知方式
				if thisUser.NotificationType == NOTIFICATION_TYPE_EMAIL {
					job.NotificationType = NOTIFICATION_TYPE_EMAIL
					job.Addr = thisUser.Email
					job.Title = parseEmailTemplate(job.Title, thisUser, class, act)
					job.Body = parseEmailTemplate(job.Body, thisUser, class, act)
				} else if thisUser.NotificationType == NOTIFICATION_TYPE_WECHAT {
					//微信
					job.NotificationType = NOTIFICATION_TYPE_WECHAT
					job.Title = parseEmailTemplate(job.Title, thisUser, class, act)
					job.Body = parseWechatBodyTitle(job.Body, thisUser, class, act, job)
					job.Addr = thisUser.WxPusherUid
				} else if thisUser.NotificationType == NOTIFICATION_TYPE_NONE {
					//已关闭通知
					continue
				}
				msgJson, err = json.Marshal(job)
				if err != nil {
					Logger.Error.Println("[定时任务][准备数据]json格式化失败", err, thisUser)
					continue
				}
				Logger.Info.Println("[定时任务][准备数据]已添加任务:", string(msgJson))
				//存入redis
				rdb.LPush(ctx, "SIGNIN_APP:NOTI_LIST", string(msgJson))

				//更新提醒次数
				err = actNotiUserTimesIncr(act, thisUser.UserID)
				if err != nil {
					Logger.Error.Println("[定时任务][准备数据]更新提醒次数失败", err)
				}
			}

			//活动参与率未达标，群发给管理员
			if sts.Done < sts.Total {
				timeHour := time.Now().Format("15")
				et, err := strconv.ParseInt(act.EndTime, 10, 64)
				if err != nil {
					Logger.Error.Println("[定时任务][准备数据][群发给管理员]时间转换失败")
				} else {
					if (timeHour == "12" && et-time.Now().Unix() < 6*60*60) || (timeHour == "18" && et-time.Now().Unix() < 18*60*60) {
						err = ActEndingBulkSend(classID, act)
					}
					if err != nil {
						Logger.Error.Println("[定时任务][准备数据][群发给管理员]发送报错：", err)
					}

					//给用户推送站内警告
					for i := range sts.UnfinishedList {
						thisUser := userMap[sts.UnfinishedList[i].UserID]
						noti, err := makeActInnerNoti(actID, thisUser.UserID, ACT_NOTI_TYPE_TIME_WARN)
						err = pushInnerNoti(thisUser.UserID, noti)
						if err != nil {
							Logger.Error.Println("[定时任务][准备数据]给用户推送站内警告失败：", err)
							continue
						}
						Logger.Info.Printf("[定时任务][准备数据]已给%s推送站内警告:%s\n", thisUser.Name, noti.Text)
					}
				}
			}
		}

	}

}

func SendDailyNotification() {
	if config.General.Production == false {
		time.Sleep(5 * time.Second)
	}
	jobJsons, err := rdb.LRange(ctx, "SIGNIN_APP:NOTI_LIST", int64(0), int64(-1)).Result()
	if err != nil {
		Logger.Error.Println("[定时任务][发送]读取队列内容失败", err)
		return
	}
	if len(jobJsons) == 0 {
		//无发送任务
		return
	}

	total := len(jobJsons)
	for i := range jobJsons {
		Logger.Info.Println("[定时任务][发送][", i+1, "/", total, "]已读取", jobJsons[i])
		job := new(NotifyJob)
		err = json.Unmarshal([]byte(jobJsons[i]), job)
		if err != nil {
			Logger.Error.Println("[定时任务][发送][", i+1, "/", total, "]json解析失败", err)
			continue
		}
		if job.NotificationType == NOTIFICATION_TYPE_EMAIL {
			mailTask, err := newMailTask(job.Addr, job.Title, job.Body)
			if err != nil {
				Logger.Error.Println("[定时任务][发送][", i+1, "/", total, "]创建邮件任务失败", err)
				continue
			}
			MailQueue <- mailTask
		} else if job.NotificationType == NOTIFICATION_TYPE_WECHAT {
			//微信
			WechatQueue <- job
		}
		Logger.Info.Println("[定时任务][发送][", i+1, "/", total, "]发送成功")
	}
	rdb.Del(ctx, "SIGNIN_APP:NOTI_LIST")
}

func CleanExpiredFiles() {
	nowTS := time.Now().Unix()
	files := make([]dbFile, 0)
	err := db.Select(&files, "select * from `file` where `status`!=?", FILE_STATUS_DELETED)
	if err != nil {
		Logger.Error.Println("[清理过期文件][失败] - 从db获取文件失败：", err)
		return
	}
	for i := range files {
		expTime, err := strconv.ParseInt(files[i].ExpTime, 10, 64)
		if err != nil {
			Logger.Error.Printf("[清理过期文件][失败] - 文件[%s]exp_time转换失败：%s", files[i].FileName, err.Error())
			continue
		}
		if expTime > nowTS { //未过期
			continue
		}

		if files[i].Status == FILE_STATUS_LOCAL {
			//删除本地文件
			go trashCleaner(files[i].Local, 0)
		} else if files[i].Status == FILE_STATUS_REMOTE {
			//删除远端文件
			err = cosFileDel(files[i].Remote)
			if err != nil {
				Logger.Error.Printf("[清理过期文件][失败] - 文件[%s]删除失败：%s", files[i].FileName, err.Error())
				continue
			}
		} else { //异常
			Logger.Error.Printf("[清理过期文件][失败] - 文件状态异常%s", files[i].Status)
			continue
		}

		//更新数据库
		_, err = db.Exec("update `file` set `status` = ? where `file_id`=?", FILE_STATUS_DELETED, files[i].FileID)
		if err != nil {
			Logger.Error.Printf("[清理过期文件][失败] - [%s]数据库更新失败：%s", files[i].FileName, err.Error())
			continue
		}
		Logger.Info.Printf("[清理过期文件][成功] - 已成功删除%s", files[i].FileName)
	}
}

//随机选取发送模板
func getTemplate(msgType int, level int) (*NotifyJob, error) {
	job := new(NotifyJob)
	tpls := make([]dbTplItem, 0)
	err := db.Select(&tpls, "select `title`,`body` from `msg_template` where `msg_type`=? and `level`=?", msgType, level)
	if err != nil {
		return nil, err
	}

	rand.Seed(time.Now().UnixNano())
	id := rand.Intn(len(tpls))
	job.Title = tpls[id].Title
	job.Body = tpls[id].Body
	return job, err
}

func trashCleaner(fileName string, minute int) {
	Logger.Info.Println("[自动清理]已收到任务清理:", fileName)
	time.Sleep(time.Duration(minute) * time.Minute)
	err := os.RemoveAll(fileName)
	if err != nil {
		Logger.Info.Println("[自动清理]清理:", fileName, "失败:", err)
	} else {
		Logger.Info.Println("[自动清理]清理:", fileName, "成功")
	}
}
